workers/unity/Packages/com.improbable.gdk.tools/RedirectedProcess.cs

```diff
@@ -126,131 +133,96 @@ public RedirectedProcess RedirectOutputOptions(OutputRedirectBehaviour redirectB
-                    {
-                        errorString = ProcessSpatialOutput(errorString);
-                    }
-
-                    if ((outputRedirectBehaviour & OutputRedirectBehaviour.RedirectStdErr) !=
-                        OutputRedirectBehaviour.None)
-                    {
-                        Debug.LogError(errorString);
-                    }
+                    processStandardOutput.Add(output);
+                }
+            });
+            AddErrorProcessing(error =>
+            {
+                lock (processStandardOutput)
+                {
+                    processStandardError.Add(error);
+                }
+            });
 
-                    if (outputLog != null)
-                    {
-                        lock (outputLog)
-                        {
-                            outputLog.AppendLine(ProcessSpatialOutput(errorString));
-                        }
-                    }
+            if (token == null)
+            {
+                var exitCode = Run();
 
-                    foreach (var errorProcessor in errorProcessors)
-                    {
-                        try
-                        {
-                            errorProcessor.Invoke(errorString);
-                        }
-                        catch (Exception e)
-                        {
-                            Debug.LogException(e);
-                        }
-                    }
+                return new RedirectedProcessResult
+                {
+                    ExitCode = exitCode,
+                    Stdout = processStandardOutput,
+                    Stderr = processStandardError
                 };
+            }
 
-                process.EnableRaisingEvents = true;
+            var result = await RunWithCancel(token.Value);
 
-                if (!process.Start())
-                {
-                    throw new Exception(
-                        $"Failed to run {info.FileName} {info.Arguments}");
-                }
+            return new RedirectedProcessResult
+            {
+                ExitCode = result,
+                Stdout = processStandardOutput,
+                Stderr = processStandardError
+            };
+        }
 
-                process.BeginOutputReadLine();
-                process.BeginErrorReadLine();
+        private Task<int> RunWithCancel(CancellationToken token)
+        {
+            var (process, outputLog) = SetupProcess();
 
-                process.WaitForExit();
+            var taskCompletionSource = new TaskCompletionSource<int>();
 
-                if (outputLog == null)
-                {
-                    return process.ExitCode;
-                }
+            // Register a handler to process the output when it exits.
+            process.Exited += (sender, args) =>
+            {
+                taskCompletionSource.SetResult(process.ExitCode);
```

Comment:
What happens in here when the token is cancelled and the process exits due to `Kill()`?

Reply:
The Exited event should fire and I would _hope_ you get a non-zero exit code. I can test this however.

I also realized this code may be race-y: 

- You can try to kill the process after it ends which will throw.
- The process could end before the `Exited` handler is set which may result in it never being thrown.

The former is easier to fix that the latter which will require some structural changes.

Reply:
The latter I'm not sure is an issue - may need a bit of research

Reply:
`process.Start()` was called in `SetupProcess()` which means it could immediately exit before we register the handler? Either way re-ordering these fixes the issue